core/trino-main/src/main/java/io/trino/operator/ScanFilterAndProjectOperator.java

```diff
@@ -255,7 +263,10 @@ else if (cursor != null) {
             }
 
             if (source instanceof RecordPageSource) {
-                cursor = ((RecordPageSource) source).getCursor();
+                synchronized (this) {
```

Comment:
There can be a race condition between `cancel` and cursor assignment. We should cancel here immediately if `cancel` was set  beforehand